### ISSUE TITLE
Validate coinbase miner subsidy and miner fees

### ISSUE THREAD

**Issue by yaahc** [ZcashFoundation/zebra#1162]
## Motivation

We should check Coinbase transactions, to make sure they satisfy the miner subsidy and miner fee consensus rules.

### Specifications

> The total value in zatoshi of transparent outputs from a coinbase transaction, minus vbalanceSapling, minus vbalanceOrchard, MUST NOT be greater than the value in zatoshi of miner subsidy plus the transaction fees paid by transactions in this block .

https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus

#### Block Subsidy and Funding Streams 

Implemented in #338:

> Like Bitcoin, Zcash creates currency when blocks are mined. The value created on mining a block is called the block subsidy. 
>
> [Canopy onward] The block subsidy is composed of a miner subsidy and a series of funding streams. 
>
> As in Bitcoin, the miner of a block also receives transaction fees.

https://zips.z.cash/protocol/protocol.pdf#subsidyconcepts

#### Transaction value balances

Implemented in #2381:

> Transparent inputs to a transaction insert value into a transparent transaction value pool associated with the transaction, and transparent outputs remove value from this pool. As in Bitcoin, the remaining value in the pool is available to miners as a fee.
> 
> Consensus rule: The remaining value in the transparent transaction value pool MUST be nonnegative.

Shielded value balances also add or remove funds from the transparent value pool. For the full set of shielded value balance rules, see ZIP-209: Validate transaction value balances #2381.

**Comment by teor2345** [comment]
I updated the modules in this ticket based on @hdevalence's comments.

@oxarbitrage can you add a `subsidy::coinbase_script` module to the block subsidy design RFC?

The `subsidy::coinbase_script` module should use the `zcash_script` crate to verify the consensus rules for transparent founders reward and funding stream outputs.

**Comment by teor2345** [comment]
Updated the modules again based on today's Zebra meeting.

**Comment by teor2345** [comment]
Soft-blocked by #1166.

We could do this work before splitting out the `CoinbaseVerifier`, but it would be better to do #1166 first.

**Comment by teor2345** [comment]
We're no longer creating a CoinbaseVerifier in #1166, so this ticket isn't blocked any more.

**Comment by teor2345** [comment]
This ticket validates coinbase inputs, so it is out of scope for the moment.
